internal/jobtracker/jobtracker.go

```diff
@@ -0,0 +1,134 @@
+// Copyright 2025 Canonical.
+
+// Package jobtracker provides a way to run routines in one instance of JIMM and track them in another.
+// That is, their status can be checked and they can be stopped.
+package jobtracker
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/juju/zaputil/zapctx"
+	"go.uber.org/zap"
+)
+
+// Store defines the interface for tracking the lifecycle and status of jobs.
+// It provides methods to add a new job, update its status (running, successful, or failed),
+// and retrieve a stop signal for a specific job.
+type Store interface {
+	AddJob(ctx context.Context, jobType string) (uuid.UUID, error)
+	SetJobRunning(ctx context.Context, jobId uuid.UUID) error
+	SetJobSuccessful(ctx context.Context, jobId uuid.UUID) error
+	SetJobFailed(ctx context.Context, jobId uuid.UUID, jobErr error) error
+	GetJobStopSignal(ctx context.Context, jobId uuid.UUID) (stopSignal bool, err error)
+}
+
+// Tracker manages job tracking operations using a provided JobTrackerStore.
+// It periodically performs tasks based on the specified stopInterval duration.
+type Tracker struct {
+	store        Store
+	stopInterval time.Duration
+}
+
+// NewJobTracker creates and returns a new Tracker instance using the provided JobTrackerStore and stopInterval.
+// It returns an error if the store is nil or if stopInterval is not greater than zero.
+func NewJobTracker(store Store, stopInterval time.Duration) (*Tracker, error) {
+	if store == nil {
+		return nil, errors.New("store cannot be nil")
+	}
+	if stopInterval <= 0 {
+		return nil, errors.New("stopInterval must be greater than zero")
+	}
+
+	return &Tracker{
+		stopInterval: stopInterval,
+		store:        store,
+	}, nil
+}
+
+// Run runs a new job and returns the job ID.
+func (j *Tracker) Run(ctx context.Context, jobType string, job func(ctx context.Context) error, maxDuration time.Duration) (uuid.UUID, error) {
+	jobId, err := j.store.AddJob(ctx, jobType)
+	if err != nil {
+		return jobId, err
+	}
+
+	go j.handleJob(ctx, jobId, maxDuration, job)
```

Comment:
You need to use a new context here. Or better yet, don't pass a context in, create one inside handleJob().
Otherwise the parent context can be cancelled and unintentionally cancel the job.